### ISSUE TITLE
[bug #2419] Disabling the Watchdog

### ISSUE THREAD

**Issue by avrs-admin** [avrdudes/avr-libc#16]
Tue 28 Jan 2003 11:34:23 PM CET

The latest (and prior versions?) compiler generates the following code to disable the watchdog --- wdt_disable()---:

Outside interrupt routines:

in **_tmp_reg_**,__SREG__
cli
wdr
out 33,r24
out **_SREG_**,__tmp_reg__
out 33,r18

// 33 is the WDTCR

Inside interrupt routines:

ldi r24,lo8(24)
ldi r25,hi8(24)
out 33, r24
out 33, **_zero_reg_**

If the user is using any watchdog timeout other than 17ms, the interrupt version of the watchdog disable may actually cause a watchdog reset! To prevent this, the WDR instruction should be added in front of the interrupt code as in the non-interrupt code, or the prescaler bits should be preserved during the disable operation. In 'C' the user can avoid the problem by issuing a wdt_reset() instruction just before the wdt_disable() instruction.

In other words, any time you wish to disable the watchdog inside an interrupt routine use the following:

wdt_reset();
wdt_disable();
--

This issue was migrated from https://savannah.nongnu.org/bugs/?2419

**Comment by avrs-admin** [comment]
Joerg Wunsch <joerg_wunsch>
Wed 29 Jan 2003 11:02:06 AM CET

I can't seem to find how this should happen at all.
avr/wdt.h only contains inline asm that doesn't know
whether it's going to be compiled inside an interrupt
context or not.

Should this be a real bug, i'll handle it anyway.

**Comment by avrs-admin** [comment]
Joerg Wunsch <joerg_wunsch>
Thu 30 Jan 2003 06:19:02 PM CET

Problem was caused by a redefined wdt_disable() macro.
Strange only that gcc didn't warn about the redefinition.
